File: DynamicKey/AgoraDynamicKey/ruby/src/dynamic_key5.rb
# coding: utf-8
require 'openssl'
require 'base64'
module DynamicKey5
  module_function

# service type
  MEDIA_CHANNEL_SERVICE = 1
  RECORDING_SERVICE = 2
  PUBLIC_SHARING_SERVICE = 3
  IN_CHANNEL_PERMISSION = 4

# extra key
  ALLOW_UPLOAD_IN_CHANNEL = 1

# permision
  NO_UPLOAD = "0"
  AUDIO_VIDEO_UPLOAD= "3"

  def gen_public_sharing_key(
      app_id,
      app_certificate,
      channel_name,
      unix_ts,
      random_int,
      uid,
      expired_ts)
    gen_dynamic_key(
        PUBLIC_SHARING_SERVICE,
        app_id,
        app_certificate,
        channel_name,
        unix_ts,
        random_int,
        uid,
        expired_ts, {})
  end


  def gen_recording_key(
      app_id,
      app_certificate,
      channel_name,
      unix_ts,
      random_int,
      uid,
      expired_ts)

    gen_dynamic_key(
        RECORDING_SERVICE,
        app_id,
        app_certificate,
        channel_name,
        unix_ts,
        random_int,
        uid,
        expired_ts,
        {})
  end

  def gen_media_channel_key(
      app_id,
      app_certificate,
      channel_name,
      unix_ts,
      random_int,
      uid,
      expired_ts)

    gen_dynamic_key(
        MEDIA_CHANNEL_SERVICE,
        app_id,
        app_certificate,
        channel_name,
        unix_ts,
        random_int,
        uid,
        expired_ts,
        {})
  end

  def gen_in_channel_permission_key(
      app_id,
      app_certificate,
      channel_name,
      unix_ts,
      random_int,
      uid,
      expired_ts,
      permission)

    extra={}
    extra[ALLOW_UPLOAD_IN_CHANNEL]=permission
    gen_dynamic_key(
        IN_CHANNEL_PERMISSION,
        app_id,
        app_certificate,
        channel_name,
        unix_ts,
        random_int,
        uid,
        expired_ts,
        extra)
  end

  def gen_dynamic_key(
      service_type,
      app_id,
      app_certificate,
      channel_name,
      unix_ts,
      random_int,
      uid,
      expired_ts,
      extra
  )
    uid = uid & 0xFFFFFFFF
    random_int = random_int & 0xFFFFFFFF
    signature = gen_signature(
        service_type,
        app_id,
        app_certificate,
        channel_name,
        unix_ts,
        random_int,
        uid,
        expired_ts,
        extra
    )
    version = sprintf('%03d', 5)
    content = pack_uint16(service_type)
    content+= pack_string(signature)
    content+= pack_string([app_id].pack("H*"))
    content+= pack_uint32(unix_ts)
    content+= pack_uint32(random_int)
    content+= pack_uint32(expired_ts)
    content+= pack_map(extra)

    version + Base64.encode64(content).gsub(/\n/,'')
  end

  def gen_signature(
      service_type,
      app_id,
      app_certificate,
      channel_name,
      unix_ts,
      random_int,
      uid,
      expired_ts,
      extra)
    raw_app_certificate = [app_certificate].pack("H*")
    raw_app_id = [app_id].pack("H*")

    content = pack_uint16(service_type)
    content += pack_string(raw_app_id)
    content += pack_uint32(unix_ts)
    content += pack_int32(random_int)
    content += pack_string(channel_name)
    content += pack_uint32(uid)
    content += pack_uint32(expired_ts)
    content += pack_map(extra)

    OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new("sha1"), raw_app_certificate, content).upcase
  end

  def pack_uint16(x)
    [x].pack('<S')
  end

  def pack_uint32(x)
    [x].pack('<I')
  end

  def pack_int32(x)
    [x].pack('<i')
  end

  def pack_string(string)
    pack_uint16(string.length) + string
  end

  def pack_map(m)
    ret = pack_uint16(m.size)
    m.each do |k, v|
      pack_uint16(k) + pack_string(v.to_s)
    end
    ret
  end
end

File: DynamicKey/AgoraDynamicKey/nodejs/src/DynamicKey5.js
var crypto = require('crypto');

var version = "005";
var noUpload = "0";
var audioVideoUpload = "3";

var generatePublicSharingKey = function (appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs) {
    channelName=channelName.toString();
    return generateDynamicKey(appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs, null, PUBLIC_SHARING_SERVICE);
};

var generateRecordingKey = function (appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs) {
    channelName=channelName.toString();
    return generateDynamicKey(appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs, null, RECORDING_SERVICE);
};

var generateMediaChannelKey = function (appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs) {
    channelName=channelName.toString();
    return generateDynamicKey(appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs, null, MEDIA_CHANNEL_SERVICE);
};

var generateInChannelPermissionKey = function (appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs, permission) {
    var extra = {};
    extra[ALLOW_UPLOAD_IN_CHANNEL] = permission;
    return generateDynamicKey(appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs, extra, IN_CHANNEL_PERMISSION);
};

var generateDynamicKey = function (appID, appCertificate, channelName, unixTs, randomInt, uid, expiredTs, extra, serviceType) {
    var signature = generateSignature5(appCertificate, serviceType, appID, unixTs, randomInt, channelName, uid, expiredTs, extra);
    var content = DynamicKey5Content({
        serviceType: serviceType
        , signature: signature
        , appID: hexDecode(appID)
        , unixTs: unixTs
        , salt: randomInt
        , expiredTs: expiredTs
        , extra: extra}).pack();
    return version + content.toString('base64');
};

module.exports.version = version;
module.exports.noUpload = noUpload;
module.exports.audioVideoUpload = audioVideoUpload;
module.exports.generatePublicSharingKey = generatePublicSharingKey;
module.exports.generateRecordingKey = generateRecordingKey;
module.exports.generateMediaChannelKey = generateMediaChannelKey;
module.exports.generateInChannelPermissionKey = generateInChannelPermissionKey;
module.exports.generateDynamicKey = generateDynamicKey;

var generateSignature5 = function(appCertificate, serviceType, appID, unixTs, randomInt, channelName, uid, expiredTs, extra) {
    // decode hex to avoid case problem
    var rawAppID = hexDecode(appID);
    var rawAppCertificate = hexDecode(appCertificate);

    var m = Message({
        serviceType: serviceType
        , appID: rawAppID
        , unixTs: unixTs
        , salt: randomInt
        , channelName: channelName
        , uid: uid
        , expiredTs: expiredTs
        , extra: extra
    });

    var toSign = m.pack();
    return encodeHMac(rawAppCertificate, toSign);
};

var encodeHMac = function(key, message) {
    return crypto.createHmac('sha1', key).update(message).digest('hex').toUpperCase();
};

var hexDecode = function(str) {
    return Buffer.from(str, 'hex');
};

var ByteBuf = function() {
    var that = {
        buffer: Buffer.alloc(1024)
        , position: 0
    };

    that.buffer.fill(0);

    that.pack = function() {
        var out = Buffer.alloc(that.position);
        that.buffer.copy(out, 0, 0, out.length);
        return out;
    };

    that.putUint16 = function(v) {
        that.buffer.writeUInt16LE(v, that.position);
        that.position += 2;
        return that;
    };

    that.putUint32 = function(v) {
        that.buffer.writeUInt32LE(v, that.position);
        that.position += 4;
        return that;
    };

    that.putBytes = function(bytes) {
        that.putUint16(bytes.length);
        bytes.copy(that.buffer, that.position);
        that.position += bytes.length;
        return that;
    };

    that.putString = function(str) {
        return that.putBytes(Buffer.from(str));
    };

    that.putTreeMap = function(map) {
        if (!map) {
            that.putUint16(0);
            return that;
        }

        that.putUint16(Object.keys(map).length);
        for (var key in map) {
            that.putUint16(key);
            that.putString(map[key]);
        }

        return that;
    };

    return that;
}

var DynamicKey5Content = function(options) {
    options.pack = function() {
        var out = ByteBuf();
        return out.putUint16(options.serviceType)
            .putString(options.signature)
            .putBytes(options.appID)
            .putUint32(options.unixTs)
            .putUint32(options.salt)
            .putUint32(options.expiredTs)
            .putTreeMap(options.extra)
            .pack();
    }

    return options;
}

var Message = function(options) {
    options.pack = function() {
        var out = ByteBuf();
        return out.putUint16(options.serviceType)
            .putBytes(options.appID)
            .putUint32(options.unixTs)
            .putUint32(options.salt)
            .putString(options.channelName)
            .putUint32(options.uid)
            .putUint32(options.expiredTs)
            .putTreeMap(options.extra)
            .pack();
    }

    return options;
}

// InChannelPermissionKey
var ALLOW_UPLOAD_IN_CHANNEL = 1;

// Service Type
var MEDIA_CHANNEL_SERVICE = 1;
var RECORDING_SERVICE = 2;
var PUBLIC_SHARING_SERVICE = 3;
var IN_CHANNEL_PERMISSION = 4;
